patches/server/0739-Handle-bed-explosion-correctly.patch

```diff
@@ -0,0 +1,33 @@
+index 82eaf3bd7332b173197d81eaebdf58c6f43d6a23..bf3bcdcd1a7851c7dfe83ce6acda5b8d5f7ef152 100644
+--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
++++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
+@@ -1253,7 +1253,7 @@ public class ServerPlayer extends Player {
+     // CraftBukkit start - moved bed result checks from below into separate method
+     private Either<Player.BedSleepingProblem, Unit> getBedResult(BlockPos blockposition, Direction enumdirection) {
+         if (!this.isSleeping() && this.isAlive()) {
+-            if (!this.level.dimensionType().natural()) {
++            if (!this.level.dimensionType().natural() || !this.level.dimensionType().bedWorks()) { // Paper add not bedWorks to NOT_POSSIBLE_HERE for handle explode
+                 return Either.left(Player.BedSleepingProblem.NOT_POSSIBLE_HERE);
+             } else if (!this.bedInRange(blockposition, enumdirection)) {
+                 return Either.left(Player.BedSleepingProblem.TOO_FAR_AWAY);
+diff --git a/src/main/java/net/minecraft/world/level/block/BedBlock.java b/src/main/java/net/minecraft/world/level/block/BedBlock.java
+index 163a7861f987c3832aac51cc6df950c768546731..4139567dc1514a27394a0b35d697668468aebc13 100644
+--- a/src/main/java/net/minecraft/world/level/block/BedBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/BedBlock.java
+@@ -113,7 +113,8 @@ public class BedBlock extends HorizontalDirectionalBlock implements EntityBlock
+                 player.startSleepInBed(pos).ifLeft((entityhuman_enumbedresult) -> {
+                     // Paper start - PlayerBedFailEnterEvent
+                     if (entityhuman_enumbedresult != null) {
+-                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = new io.papermc.paper.event.player.PlayerBedFailEnterEvent((org.bukkit.entity.Player) player.getBukkitEntity(), io.papermc.paper.event.player.PlayerBedFailEnterEvent.FailReason.VALUES[entityhuman_enumbedresult.ordinal()], org.bukkit.craftbukkit.block.CraftBlock.at(world, finalblockposition), entityhuman_enumbedresult == Player.BedSleepingProblem.NOT_POSSIBLE_HERE, io.papermc.paper.adventure.PaperAdventure.asAdventure(entityhuman_enumbedresult.getMessage()));
++                        boolean bedWillExplode = entityhuman_enumbedresult == Player.BedSleepingProblem.NOT_POSSIBLE_HERE && !world.dimensionType().bedWorks(); // Paper Handle explode flag
++                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = new io.papermc.paper.event.player.PlayerBedFailEnterEvent((org.bukkit.entity.Player) player.getBukkitEntity(), io.papermc.paper.event.player.PlayerBedFailEnterEvent.FailReason.VALUES[entityhuman_enumbedresult.ordinal()], org.bukkit.craftbukkit.block.CraftBlock.at(world, finalblockposition), bedWillExplode, io.papermc.paper.adventure.PaperAdventure.asAdventure(entityhuman_enumbedresult.getMessage()));
```

Comment:
I don't see why this diff is needed, isnt `entityhuman_enumbedresult` already taking into account if the dimension supports beds. I thought that's what the other big diff was for.

Reply:
entityhuman_enumbedresult return NOT_POSSIBLE_HERE for two cases

- If the dimension is not natural (like nether o end)
- If the dimention has set bed in not works (added in this PR)

Then need specify the correct logic for bed explode (later in the paper event the user can change this like always) because if dimension is not natural only need affect in this case for the bed use not for the explode.

Based in the short-cirtuit of craftbukkit to the validation where check if bed can explode and the wiki.
![image](https://user-images.githubusercontent.com/3602279/128544399-19da4a69-4ccd-4804-9299-6a9cc483d41b.png)

Reply:
Oh I see, the bedWorks flag is what determines the blowing up, not the `natural` thing.

This is a large-ish behavioral change that might break stuff, idk how long its been this way, but I tend to be in favor of breaking more stuff than most people so I think its fine.

Do you even need to check `entityhuman_enumbedresult` then? It should always be `NOT_POSSIBLE_HERE` if `!world.dimensionType().bedWorks()` is true.

Reply:
Its based in my PR from craftbukkit where PlayerBedEnterEvent can change the result again. then if change the result to ALLOW for example its because i think the dev want avoid the bed explode and want the player sleep in teory (they can use BlockExplosionEvent for this too)

Reply:
I don't think you can change the bed result to NOT_POSSIBLE_HERE from PlayerBedEnterEvent.

Reply:
Checking in Craftbukkit code is rigth.. if result its set to DENY the result is OTHER_PROBLEM then only need check the bedworks... 
Then in craftbukkit the logic if very strange...

Changed the line to only check for bedWorks.